c10/cuda/CUDACachingAllocator.cpp

```diff
@@ -1193,14 +1270,250 @@ class DeviceCachingAllocator {
+   * will not free its blocks to the main caching allocator until cuda graph use
+   * is finished to prevent an allocation from eager clobbering the memory from
+   * a live but unaccounted for tensor that was created during replay.
+   *
+   * `make_graphed_callables`, a series of separate callables chained in
+   * successive cuda graphs, can share a memory pool because after a cuda graph
+   * recording the allocations in the shared private pool exactly reflect the
+   * tensors that are allocated.
+   *
+   * We would like to extend callable chaining to support a graphed callable
+   * tree. In this scenario, we have a tree of callable chains which will be
+   * captured with cuda graphs. In the diagram below, we have a tree with four
+   * callables, A, B, C, and D. Suppose we have captured, and subsequently
+   * replayed, A, B, and C. Then on a new invocation, we replay A and B, but
+   * would now like to record D. At this point the private pool will not reflect
+   * any of the live tensors created during graph replay. Allocations made
+   * during a new recording with the pool could overwrite those live tensors.
+   *
+   * In order to record a new graph capture after replaying prior callables in
+   * the tree, we need the allocator to reflect the state of the live tensors.
+   * We checkpoint the state of the private pool after each recording, and then
+   * reapply it when we are starting a new recording chain. Additionally, we
+   * must free the allocations for any tensors that died between the end of our
+   * previous graph replaying and our new recording (TODO). All of the allocated
+   * segments that existed in the checkpointed state must still exist in the
+   * pool. There may also exist new segments, which we will free (TODO : link
+   * note [live tensors between iterations] when it exists).
+   *
+   *
+   *  ---------------> A ---------------> B ---------------> C
+   *                                      |
+   *                                      |
+   *                                      |
+   *                                      |
+   *                                      ╰ ---------------> D
+   */
+  void setCheckpointPoolState(PrivatePoolState& pps) {
+    // To reset the caching allocator state we will
+    // - Free all the blocks currently allocated to the pool (see [live tensors
+    // between iterations])
+    // - Allocate all the blocks in a checkpointed segment, whether they are
+    // live or not
+    // - Free the blocks in a checkpointed segment which are not live
+    // This could be optimized, but it nicely reuses exiting apis, and this
+    // is not on the hot path.
+
+    // following `done outside the lock because we don't know what locks the
+    // recorder needs to have...`
+
+    CreateContextFn context_recorder = context_recorder_.load();
+    std::shared_ptr<Context> context =
+        context_recorder ? context_recorder() : nullptr;
+
+    // TODO - lock_guard vs unique_guard ??
```

Comment:
TODO

Reply:
NVM - I dont actually think this is an issue.